dali/python/nvidia/dali/external_source.py

```diff
@@ -129,18 +137,30 @@ def _is_generator_function(x):
     call = getattr(x, "__call__", None)
     return _is_generator_function(call)
 
+def _cycle_enabled(cycle):
+    if cycle is None:
+        return False
+    if cycle == False or cycle == "no":
+        return False
+    if cycle == True or cycle == "quiet" or cycle == "raise":
+        return True
+    raise ValueError("""Invalid value {} for the argument `cycle`. Valid values are
+  - "no", False or None - cycling disabled
+  - "quiet", True - quietly rewind the data
+  - "raise" - raise StopIteration on each rewind.""".format(cycle))
```

Comment:
How about making this an `assert False, ...`? This function might (and actually is...) used in try/except, so the exception might be improperly caught

Reply:
It's a different exception (the one caught is `TypeError`) - and `assert` suggests an internal error rather than argument validation failure.

Reply:
Tested with "rise" string:
```
======================================================================
ERROR: test_external_source_numpy.test_external_source_collection_cycling_raise
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/michal/.local/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/michal/.local/lib/python3.6/site-packages/nose/util.py", line 620, in newfunc
    return func(*arg, **kw)
  File "/home/michal/Projects/DALI/dali/test/python/test_external_source_impl.py", line 358, in test_external_source_collection_cycling_raise
    pipe.set_outputs(fn.external_source(batches, cycle = "rise"), fn.external_source(batch_gen, cycle = "raise"))
  File "/home/michal/.local/lib/python3.6/site-packages/nvidia/dali/external_source.py", line 495, in external_source
    use_copy_kernel = use_copy_kernel, batch = batch, **kwargs)
  File "/home/michal/.local/lib/python3.6/site-packages/nvidia/dali/external_source.py", line 340, in __init__
    callback = _get_callback_from_source(source, cycle)
  File "/home/michal/.local/lib/python3.6/site-packages/nvidia/dali/external_source.py", line 156, in _get_callback_from_source
    if _cycle_enabled(cycle):
  File "/home/michal/.local/lib/python3.6/site-packages/nvidia/dali/external_source.py", line 150, in _cycle_enabled
    - "raise" - raise StopIteration on each rewind.""".format(cycle))
ValueError: Invalid value rise for the argument `cycle`. Valid values are
  - "no", False or None - cycling disabled
  - "quiet", True - quietly rewind the data
  - "raise" - raise StopIteration on each rewind.
```